### PR TITLE
Update ERN platform to support RN 0.68.7

### PR DESCRIPTION
**Pull request by Karthiccc23**
- Update Android Plugins and tools to support RN 0.68.7. - https://github.com/electrode-io/react-native/tree/0.68-stable
    ANDROID_GRADLE_PLUGIN_VERSION = '7.0.4';
    BUILD_TOOLS_VERSION = '31.0.0';
    COMPILE_SDK_VERSION = '31';
    GRADLE_DISTRIBUTION_VERSION = '7.3.3';
    KOTLIN_VERSION = '1.6.10';
    MIN_SDK_VERSION_PRE_RN64 = '19';
    MIN_SDK_VERSION_POST_RN64 = '21';
    TARGET_SDK_VERSION = '30';

- Update RN Codegen based on react-native version
  -  gte 0.68.4 => codegen ^0.0.18 
  -  gte 0.68.2 => codegen ^0.0.17
  -  gte 0.68.0 => codegen ^0.0.13

- React Native 0.68.0 Breaking Changes - Removed fallbackResource from RCTBundleURLProvider API ([0912ee179c](https://github.com/facebook/react-native/commit/0912ee179c210fb6b2ed9afbb3f2fbc5fb8a2bb3))
-   ern run-ios Build Failed due to yargs package not found in generate-specs-cli script. Added yargs since this package is missing from generate-specs-cli script. - https://github.com/facebook/react-native/issues/36315. Issue is resolved from RN 0.72.0-rc.0 to prevent script failures - https://github.com/facebook/react-native/commit/5093e557efac35632d99cd2094527f9e01cbc519

### CODE REVIEW

File: ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.3-all.zip
+networkTimeout=10000
```
Comment (friederbluemle) [inline ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties:4]: It looks like RN 0.68.x comes with Gradle 7.3.3 (not 7.3) - Please ensure the version matches.

File: ern-container-gen-android/src/hull/gradlew [hunk 2]
```diff
@@ -80,10 +80,10 @@ do
     esac
 done
 
-APP_HOME=$( cd "${APP_HOME:-./}" && pwd -P ) || exit
-
-APP_NAME="Gradle"
+# This is normally unused
+# shellcheck disable=SC2034
 APP_BASE_NAME=${0##*/}
+APP_HOME=$( cd "${APP_HOME:-./}" && pwd -P ) || exit
```
Comment (friederbluemle) [inline ern-container-gen-android/src/hull/gradlew:86]: Do these changes possibly come from one of the later versions (8?) of Gradle? I don't see them with 7.3 (or 7.3.3) - How did you update the files?

File: ern-container-gen-ios/src/hull/ElectrodeContainer/ElectrodeReactNative.m [hunk 3]
```diff
@@ -381,7 +381,7 @@ - (void)addExtraDevMenu {
     RCTDevMenuItem *enableMinifyMenu = [RCTDevMenuItem buttonItemWithTitle:minifyTitle handler:^{
         [[RCTBundleURLProvider sharedSettings] setEnableMinification:!enableMinify];
         __strong RCTBridge *strongBridge = self.bridge;
-        strongBridge.bundleURL = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
+        strongBridge.bundleURL = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
```
Comment (friederbluemle) [inline ern-container-gen-ios/src/hull/ElectrodeContainer/ElectrodeReactNative.m:384]: Is this change related to the new architecture?

File: ern-core/src/android.ts [hunk 4]
```diff
@@ -12,14 +12,14 @@ import semver from 'semver';
 // Default value for android build config
 // ==============================================================================
 
-export const DEFAULT_ANDROID_GRADLE_PLUGIN_VERSION = '4.2.2';
+export const DEFAULT_ANDROID_GRADLE_PLUGIN_VERSION = '7.0.1';
```
Comment (friederbluemle) [inline ern-core/src/android.ts:15]: Specific reason for `7.0.1`? Looks like RN 0.68 ships with `7.0.4`.

File: ern-core/src/iosUtil.ts [hunk 5]
```diff
@@ -578,7 +578,9 @@ export const getDefaultIosDeploymentTarget = (rnVersion: string): string => {
 export const getReactNativeCodegenVersion = (
   rnVersion: string,
 ): string | null => {
-  if (semver.gte(rnVersion, '0.67.0')) {
+  if (semver.gte(rnVersion, '0.68.4')) {
+    return '^0.0.18';
+  } else if (semver.gte(rnVersion, '0.67.0')) {
```
Comment (friederbluemle) [inline ern-core/src/iosUtil.ts:583]: For completeness, can you also add the following cases:

gte `0.68.2` => codegen `^0.0.17`
gte `0.68.0` => codegen `^0.0.13`

**Review by friederbluemle — COMMENTED** [review]
Looks good overall 👍

File: ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (Karthiccc23) [reply ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties:4]: I picked up this change from changelog https://github.com/facebook/react-native/blob/main/CHANGELOG.md#v0680. But i did check the branch for 0.68 and I see 7.3.3

File: ern-container-gen-ios/src/hull/ElectrodeContainer/ElectrodeReactNative.m (hunk 3 above)
Comment (Karthiccc23) [reply ern-container-gen-ios/src/hull/ElectrodeContainer/ElectrodeReactNative.m:384]: not related to new arch. But again from changelog from 68.0 had breaking changes.

File: ern-core/src/android.ts (hunk 4 above)
Comment (Karthiccc23) [reply ern-core/src/android.ts:15]: from changelog 68.0 here as well

File: ern-core/src/iosUtil.ts (hunk 5 above)
Comment (Karthiccc23) [reply ern-core/src/iosUtil.ts:583]: sure!

File: ern-container-gen-android/src/hull/gradlew (hunk 2 above)
Comment (Karthiccc23) [reply ern-container-gen-android/src/hull/gradlew:86]: updating back to 7.3.3.

**Review by leekuo — APPROVED** [review]

File: ern-core/src/android.ts [hunk 6]
```diff
@@ -12,14 +12,14 @@ import semver from 'semver';
...
-export const DEFAULT_ANDROID_GRADLE_PLUGIN_VERSION = '4.2.2';
+export const DEFAULT_ANDROID_GRADLE_PLUGIN_VERSION = '7.0.4';
 export const DEFAULT_ANDROIDX_APPCOMPACT_VERSION = '1.1.0';
 export const DEFAULT_ANDROIDX_LIFECYCLE_EXTENSIONS_VERSION = '2.1.0';
-export const DEFAULT_BUILD_TOOLS_VERSION = '30.0.2';
-export const DEFAULT_COMPILE_SDK_VERSION = '30';
-export const DEFAULT_GRADLE_DISTRIBUTION_VERSION = '7.2';
+export const DEFAULT_BUILD_TOOLS_VERSION = '31.0.0';
+export const DEFAULT_COMPILE_SDK_VERSION = '31';
+export const DEFAULT_GRADLE_DISTRIBUTION_VERSION = '7.3';
```
[the remark this answers is not in the document]
Comment (Karthiccc23) [reply ern-core/src/android.ts:20]: sure.

File: ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties [hunk 7]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.3.3-all.zip
+networkTimeout=10000
```
[the remark this answers is not in the document]
Comment (Karthiccc23) [reply ern-container-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties:4]: yep will remove it.

File: ern-runner-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-{{{gradleDistributionVersion}}}-all.zip
```
[the remark this answers is not in the document]
Comment (Karthiccc23) [reply ern-runner-gen-android/src/hull/gradle/wrapper/gradle-wrapper.properties:3]: generated the gradle for 7.3.3  and this got overwritten. Will switch back to point to gradleDistributionVersion

**Review by friederbluemle — APPROVED** [review]
Once done, ready to merge 👍

**Comment by Karthiccc23** [comment]
> The Gradle Wrapper scripts (`gradlew`/`gradlew.bat`), as well as `gradle-wrapper.jar` still need to be updated to match version 7.3.3 - Once that's done it should be ready to merge

updated the scripts 👍
